src/System.Data.SqlClient/src/System/Data/SqlClient/SqlDataReader.cs

```diff
@@ -2583,85 +2583,136 @@ private T GetFieldValueInternal<T>(int i)
                         }
                         else
                         {
-                            // try the type cast to throw the invalid cast exception and inform the user what types they're trying to use and that why it is wrong
-                            return (T)clrValue;
+                            object clrValue = null;
+                            if (!data.IsNull)
+                            {
+                                clrValue = GetValueFromSqlBufferInternal(data, metaData);
+                            }
+                            if (clrValue is null) // covers IsNull and when there is data which is present but is a clr null somehow
+                            {
+                                return (T)(object)SqlTypeWorkarounds.SqlXmlCreateSqlXmlReader(
+                                    new MemoryStream(Array.Empty<byte>(), writable: false),
+                                    closeInput: true
+                                );
+                            }
+                            else if (clrValue.GetType() == typeof(string))
+                            {
+                                return (T)(object)SqlTypeWorkarounds.SqlXmlCreateSqlXmlReader(
+                                    new StringReader(clrValue as string),
+                                    closeInput: true
+                                );
+                            }
+                            else
+                            {
+                                // try the type cast to throw the invalid cast exception and inform the user what types they're trying to use and that why it is wrong
+                                return (T)clrValue;
+                            }
                         }
                     }
-                }
-                else
-                {
-                    try
-                    {
-                        return (T)GetValueFromSqlBufferInternal(data, metaData);
-                    }
-                    catch (InvalidCastException)
+                    else
                     {
-                        if (data.IsNull)
+                        try
                         {
-                            // If the value was actually null, then we should throw a SqlNullValue instead
-                            throw SQL.SqlNullValue();
+                            return (T)GetValueFromSqlBufferInternal(data, metaData);
                         }
-                        else
+                        catch (InvalidCastException)
                         {
-                            // Legitimate InvalidCast, rethrow
-                            throw;
+                            if (data.IsNull)
+                            {
+                                // If the value was actually null, then we should throw a SqlNullValue instead
+                                throw SQL.SqlNullValue();
+                            }
+                            else
```

Comment:
nit: don't need the `else`

Reply:
Agreed, it could just fall throw and have equivalent functionality but it's less obvious imo and this is the original layout so I stuck with it. Given that it's dealing with exceptions so perf is not a concern I didn't see a need to remove it, is there formatting guidance that applies to this case?

Reply:
I think your point about existing code makes sense